xdsl/transforms/printf_to_putchar.py

```diff
@@ -206,38 +206,49 @@ def print_digits(digits: SSAValue, absolute_value: scf.If) -> scf.For:
         # Return from itoa function
         func.Return()
 
-    return inline_itoa_func
+    return mlir_itoa_func
 
 
-if __name__ == "__main__":
-    printer = Printer()
-    printer.print(get_inline_itoa())
-
-
-class LowerPrintCharToPutcharPass(ModulePass):
-    name = "lower-printchar-to-putchar"
+class PrintfToPutcharPass(ModulePass):
+    name = "printf-to-putchar"
 
     # lower to func.call
     def apply(self, ctx: MLContext, op: ModuleOp) -> None:
-        print_int_walker = PatternRewriteWalker(
-            GreedyRewritePatternApplier(
-                [
-                    ConvertPrintIntToItoa(),
-                ]
-            ),
-            apply_recursively=True,
-        )
-        print_int_walker.rewrite_module(op)
-        # This has to happen first, since inline itoa
-        # contains some references to printf.print_char
-        op.body.block.add_ops([get_inline_itoa()])
-        # Add function implementation for inline_itoa
-        # TODO Don't insert this if the operation is not there!
-        print_char_walker = PatternRewriteWalker(
-            GreedyRewritePatternApplier([LowerPrintCharToPutchar()]),
-            apply_recursively=True,
-        )
-        print_char_walker.rewrite_module(op)
-        # Add external putchar reference
-        # TODO Don't insert this if the operation is not there!
-        op.body.block.add_ops([func.FuncOp.external("putchar", [i32], [i32])])
+        # Check if there are any printints
+        contains_printint = False
+        for op_in_module in op.walk():
+            if isinstance(op_in_module, PrintIntOp):
+                contains_printint = True
+                # One printint encountered is enough
+                break
+        if contains_printint:
+            print_int_walker = PatternRewriteWalker(
+                GreedyRewritePatternApplier(
```

Comment:
```suggestion
                ConvertPrintIntToItoa()
```
I don't thin ka greedy rewrite pattern applier with a single patern is any different to just that pattern. Also, I don't think we want to apply it recursively. Applying the pattern recursively would allow us to run the rewrite pattern on ops that it creates, but the `ConvertPrintIntToItoa` pattern does not create new `print_int` ops so it won't ever be applied in the recursive step.

Reply:
Similarly below.